libs/wire-api/src/Data/Swagger/Typed.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Data.Swagger.Typed
+  ( Schema,
+    NamedSchema,
+    ToTypedSchema (..),
+    TypedSchema (TypedSchema),
+    field,
+    field',
+    named,
+    unnamed,
+    array,
+    enum,
+    schema,
+    untypedSchema,
+  )
+where
+
+import Control.Applicative
+import Control.Lens (at, view, (.~), (?~))
+import Data.Aeson (Value)
+import Data.Proxy
+import qualified Data.Swagger as S
+import qualified Data.Swagger.Declare as S
+import qualified Data.Swagger.Internal.Schema as S
+import Imports
+
+type Declare = S.Declare (S.Definitions S.Schema)
+
+-- | A typed Swagger schema
+newtype Schema (a :: *) = Schema (Declare S.Schema)
+  deriving (Functor)
+
+-- FUTUREWORK: store enough information in 'Schema' to synthesise
+-- ToJSON and FromJSON instances
+
+data NamedSchema a = NamedSchema
+  { _nsSchema :: Declare S.NamedSchema,
+    _nsDescription :: Maybe Text
+  }
+  deriving (Functor)
+
+appendMaybe :: Maybe a -> Maybe a -> Maybe a
+appendMaybe Nothing x = x
+appendMaybe x _ = x
+
+appendNS :: S.NamedSchema -> S.NamedSchema -> S.NamedSchema
+appendNS (S.NamedSchema n1 s1) (S.NamedSchema n2 s2) =
+  S.NamedSchema (appendMaybe n1 n2) (s1 <> s2)
+
+instance Applicative NamedSchema where
+  pure _ = NamedSchema (pure (S.NamedSchema Nothing mempty)) Nothing
+  NamedSchema decl1 descr1 <*> NamedSchema decl2 descr2 =
+    NamedSchema (appendNS <$> decl1 <*> decl2) (appendMaybe descr1 descr2)
+
+instance Alternative NamedSchema where
+  empty = NamedSchema (pure (S.NamedSchema Nothing mempty)) Nothing
+  NamedSchema decl1 descr1 <|> NamedSchema decl2 descr2 =
+    NamedSchema (appendNS <$> decl1 <*> decl2) (appendMaybe descr1 descr2)
+
+runNamedSchema :: NamedSchema a -> Declare S.NamedSchema
+runNamedSchema (NamedSchema decl descr) = do
+  s <- decl
+  pure $ s & (S.schema . S.description) .~ descr
+
+instance Applicative Schema where
+  pure _ = Schema (pure mempty)
+  Schema a <*> Schema b = Schema ((<>) <$> a <*> b)
+
+instance Alternative Schema where
```

Comment:
So `Applicative` and `Alternative` have the same implementation...  do you introduce both just so you have two operator preferences so you can avoid parenthesis?

Reply:
The `Applicative` and `Alternative` instances have the same implementation, but different types. For example, the binary operation of `Applicative` is:
```haskell
(<*>) :: f (a -> b) -> f a -> f b
```
while that of `Alternative` is:
```haskell
(<|>) :: f a -> f a -> f a.
```

So the simple reason for the distinction is that the types will line up with the `Applicative` instance in some cases, and with the `Alternative` instance in others, so you need both.

The deeper reason is that they are conceptually different. A schema like:
```haskell
(,) <$> schemaA <*> schemaB
```
means that you will can have fields from `A` and `B` in the resulting object, while
```haskell
schemaA <|> schemaB
```
is only allowed to have fields from `A` *or* from `B`.

Unfortunately, I haven't found a way to actually express the latter constraint in the resulting schema, and that's why the implementations are the same. It seems this should be possible in openapi3, thought.

The difference between `Applicative` and `Alternative` also becomes important when we think about using a typed schema specification to derive `FromJSON` instances. Then they would respectively implement sequencing and alternation of the corresponding parsers.

Reply:
Thanks, this all makes perfect sense.  Maybe you can add some of this to the haddocks as well?  (I know documenting haskell code is almost non-idiomatic, but it helped me here, so it may help others.  :))

Reply:
Sure, I added my explanation from this PR to the haddocks.